### PR TITLE
Guest mode and some API fixes

### PR DESCRIPTION
**Pull request by tht7**
- fixed a bug with the JSONDecoder in ApiClient that couldn't parse the response when the server sends inconsistent date formats (especially on devices with a different locale than the default en_US_POSIX)
- made the app a bit friendlier by adding a "Guest mode" now if you don't have any accounts you'll see a generic "All" feed, and even when you do have an account you can still enter "Guest mode" from the account list (we might want this in the future if we'd want to protect those accounts with FaceID and so a person handing his phone to his friend to check out the platform could remain private and enter "Guest mode")
- Made the account variable optional everywhere
- made sure to treat all the places that need the account (like upvoting) with a nice toast message

# Checklist
- [x] I have described what this PR contains

**Choose one of the following two options:**
- - [x] This PR does not introduce major changes*
- - [ ] This PR introduces major changes, and I have consulted @buresdv, @jboi or @mormaer in the *Mlem Development Matrix room*

**Choose one of the following two options:**
- - [ ] This PR does not change the UI in any way
- - [x] This PR adds new UI elements / changes the UI, and I have attached pictures or videos of the new / changed UI elements (This PR hides certain buttons when you're in "Guest mode")

# Pull Request Information

## About this Pull Request
I've added a "Guest mode" to make the app friendlier, as well as fixed some bugs related to inconsistent data formatting from the server side

## Additional Context
While I've been a developer for many many years but I'm relatively new to Swift, I would love some feedback, I would love to help and contribute to this app even if it means I'll have to learn a bit before ;)

### CODE REVIEW

**Comment by mormaer** [comment]
Hey @tht7 👋 

Would you be able to raise a separate PR with just the date decoding changes included?

I'd like to get that in as soon as possible if you have the time 🤞 

Guest mode was raised recently and I got the impression that it wasn't to be included. Personally, I think it would be a positive addition, but there are likely some UX discussions to be had around it before decisions are made... I'll leave that open to others to weigh in on, what are your thoughts on guest mode @buresdv?

File: Mlem/API/Requests/SearchRequest.swift [hunk 1]
```diff
@@ -37,9 +37,9 @@ struct SearchRequest: APIGetRequest {
         creatorId: Int? = nil,
         limit: Int? = nil
     ) {
-        self.instanceURL = account.instanceLink
-        var queryItems: [URLQueryItem] = [
-            .init(name: "auth", value: account.accessToken),
+        self.instanceURL = account?.instanceLink ?? DefaultLemmyServer
+        let queryItems: [URLQueryItem] = [
+            .init(name: "auth", value: account?.accessToken),
```
Comment (ShadowJonathan) [inline Mlem/API/Requests/SearchRequest.swift:42]: Don't make null-tolerant versions of these APIs, make additional .init's which take only the Instance URL, so that the client a layer up can provide it that default URL if it wishes so

Else this will most likely just lead to buggy and inconsistent behaviour

File: Mlem/API/APIClient.swift [hunk 2]
```diff
@@ -41,6 +43,12 @@ class APIClient {
             throw APIClientError.response(apiError, statusCode)
         }
         
+        decoder.dateDecodingStrategyFormatters = [ DateFormatter.expandedTWithTimezone,
+                                                   DateFormatter.expandedT,
+                                                   DateFormatter.standardT,
+                                                   DateFormatter.standard,
+                                                   DateFormatter.yearMonthDay ]
+
```
Comment (ShadowJonathan) [inline Mlem/API/APIClient.swift:51]: I don't know if this PR will be merged quickly, but this might be useful on its own, could you separate this in its own PR?

**Review by ShadowJonathan — CHANGES_REQUESTED** [review]
We don't want logic like this on the API layer, we want that as declarative and clear as possible.

Also, like @mormaer said, we recently have gotten the request for guest mode as well, but at this stage, we're not so sure

File: Mlem/API/APIClient.swift (hunk 2 above)
Comment (tht7) [reply Mlem/API/APIClient.swift:51]: I have! https://github.com/buresdv/Mlem/pull/208

File: Mlem/API/Requests/SearchRequest.swift (hunk 1 above)
Comment (tht7) [reply Mlem/API/Requests/SearchRequest.swift:42]: I wanted to leave the UI layer blissfuly unaware when the account is nil and having the API layer have a sensible fallback seemed like the way to go, 
I'm not really sure how to implement this cleanly in the UI layer without adding this parameter, and passing it down (with a fallback in case the account is nil) on every request, which to me looks dirty, but if you think it's the way I would be happy to change it 😄

**Comment by JakeShirley** [comment]
I can see a world where there is a "Discover a server" experience in the app.  Lemmy is definitely less than user friendly.

**Comment by buresdv** [comment]
I can see a "guest mode" working is if we choose a specific set of instances (further discussion is warranted to choose which instances they will be), give them some context (for example "this is a homepage for tech nerds", "this is a homepage for general discussion", etc.), and then let the user pick from them. We should make it discoverable to the average user without them having to scour for instances.

Before a guest mode is implemented, we also have to implement a registration workflow so the users can sign up to their chosen instance right from the app.

**Comment by ShadowJonathan** [comment]
Random; We could probably also hook this into the signup flow, i.e. now have a user scour for instances but allow them to pick from a list, which new instance to sign up at.

**Comment by mormaer** [comment]
I've created an issue for discussion to continue on this here -> #216

I think for now we may not be able to proceed with this PR until we spend some time ironing out what discovery/guest mode would look like

**Comment by mormaer** [comment]
@tht7 - will we close this PR for now until we decide on the direction we want to go in for guest mode / discovery mode?
